### PR TITLE
Refactor Database Locks code

### PR DESCRIPTION
**Pull request by Kek5chen**
The lock code with the multiple transitional types are just plain bad.
I'm taking another look at it so we have less issues in the future.

### CODE REVIEW

**Comment by Kek5chen** [comment]
The change introduced in bcdcc91e2221ef8ec513168adbcacffb71e2917a is a bit controversial.

Changing the schema file on its own, while being a valid change as far as i am aware, is kind of sketchy. We have to keep notice of any future migrations / changes that might re-generate the file and un-apply the patch which would obviously lead to issues of vicky not compiling anymore.

With the implementation of 8d06aa8d8795a8a4330ab25060ea14d381869b03 we basically make any production runs, not regenerate the schema file when running migrations automatically as it happens on runtime, so we just have to worry about not fumble committing the file and introducing a change that would break it. Pipeline would catch this, but still, it does feel wrong.

I propose these solutions:
  - A build script that will always execute on build to apply the patch programatically to the schema file.
    - Pros: 
      1. Will catch any case of migrating the database and regeneration and pipeline would catch any case of not running the code before committing after migration.
      2. Will make it pretty much impossible to forget about it since it will re-run any time any change is introduced in the schema file.
    - Cons:
      1. Might run in the background here and there.
      2. Will increase build times by ~100-120 ms (compile) and ~1-2 ms (run) with my current implementation. I'm thinking about using regex instead to improve maintainability, so i'm not sure how that'll impact that.
    
  - A makefile that should be used as the only way to run external migrations as a wrapper to the diesel cli which will also apply the patch.
    - Pros:
      1. Will do migrations and automatically take care of the schema file afterwards if needed.
      2. Will only run once and apply the change smartly.
      3. Simplicity?
    - Cons:
      1. Needs to be thought of when running migrations.
      2. Weird Makefile command with env variables that should go to diesel_cli (unclear).
      3. Could be forgotten and committed to git, but pipeline would crash.
      
I also spent some time in looking up if the change can be done in the migration file so that diesel outputs a Nullable for the primary key, but apart from Sqlite having nullable primary keys by default, this does not seem to be possible even with a type like Serial being nullable on insert.

This behavior does make sense though since on Query it would always be non-null, but it's still a bit of a nuisance that this is not possible in any way since diesel just parses whatever it's given from postgres and makes that into a rust compatible format in the schema.rs.

I'm also gonna change the id's to Uuids as we discussed, but this doesn't impact any of this in any way. Just that the script has the be changed, which is a con in both solutions.

If anyone has a better idea, pls help. 

<details>
  <summary>cats for sanity</summary>
  
image
image
image
image
 
</details>

**Comment by Kek5chen** [comment]
For the script, should we rather use a cleaner regex approach or the faster contains approach?

Regex: Build: ~400ms, Run: ~30ms
Contains and Replace: Build: ~100ms, Run: ~2ms

Sources:
C&R: [build.rs](https://github.com/wobcom/vicky/blob/99fb15e6bc90e40def4f9d0da9eac244cbb1e36e/vicky/build.rs)
<details>
<summary>Regex Script</summary>

```rust
use regex::{Captures, Regex};
use std::{fs, io};
use std::fs::File;
use std::io::{BufWriter, Write};

fn main() {
    if let Err(e) = ensure_locks_id_nullable() {
        eprintln!("Couldn't update schema.rs: {e}")
    }
}

const SCHEMA_RS_PATH: &str = "src/lib/database/schema.rs";
const OLD_TYPE: &str = "Int4";
const NEW_TYPE: &str = "Nullable<Int4>";
const REPLACEX: &str = r"(diesel::table!\s*\{\n\s*locks\s*\(id\)\s*\{\n\s*id\s*->\s*)";

fn ensure_locks_id_nullable() -> io::Result<()> {
    println!("cargo::rerun-if-changed=src/lib/database/schema.rs");
    let content = fs::read_to_string(SCHEMA_RS_PATH)?;
    let regex = Regex::new(&format!(
        "{}{}", REPLACEX,
        OLD_TYPE
    ))
    .unwrap();
    

    let new_schema = regex.replace(&content, |caps: &Captures| {
        if let Some(before) = caps.get(1) {
            format!("{}{}", before.as_str(), NEW_TYPE)
        } else {
            caps.get(0).unwrap().as_str().to_string() // always present
        }
    });

    let schema = File::create(SCHEMA_RS_PATH)?;
    let mut overwriter = BufWriter::new(schema);
    overwriter.write_all(new_schema.as_bytes())?;
    Ok(())
}
```

</details>

**Comment by johannwagner** [comment]
I like the idea to fix this issue, because it removes a lot of duplicate code.
I do not lean to any possible implementation. Each implementation should provide an list with table names, where new table names can be added easily, since this might be a common pattern.
Performance-wise, even 400ms addition build time is not a big issue in the grand scheme of things.
I am willing to accept the solution, which you prefer.

File: vicky/build.rs [hunk 1]
```diff
@@ -1,3 +1,49 @@
...
+    let mut found_locks_table = false;
+    let mut only_copy = false;
+
+    for line in reader.lines() {
+        let line = line?;
+
+        if only_copy {
+            new_schema.push_str(&line);
+        } else if !found_locks_table {
+            if line.contains("locks (id) {") {
```
Comment (johannwagner) [inline vicky/build.rs:27]: We should add an list with tables, that can be easily modified.

File: vicky/build.rs [hunk 2]
```diff
@@ -1,3 +1,49 @@
...
+                found_locks_table = true;
+            }
+            new_schema.push_str(&line);
+        } else {
+            if !line.contains("id ->") {
```
Comment (johannwagner) [inline vicky/build.rs:32]: Should we support different primary keys than `id`?

Comment (Kek5chen) [reply vicky/build.rs]: ye

File: vicky/build.rs (hunk 1 above)
Comment (Kek5chen) [reply vicky/build.rs]: ye

**Pull request merged**